src/libxtp/gwbse/bse.cc

```diff
@@ -17,497 +17,582 @@
-        _bse_singlet_coefficients_AR.col(level) = (0.5 / sqrt_eval * (_bse_singlet_energies(level) * LmT - AmB) * eigenvectors.col(level)).cast<float>();
-#endif
+    CTP_LOG(ctp::logDEBUG, _log)
+        << ctp::TimeStamp() << " Lapack Diagonalization" << flush;
 
-      }
+    hstart = std::chrono::system_clock::now();
+    Eigen::MatrixXd hfull = h.get_full_matrix();
+    hend = std::chrono::system_clock::now(); 
 
-      return;
-    }
-        
-    
-        template <typename T>
-        void BSE::Add_Hqp(Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& H) {
-            
-            int offset=_opt.vmin-_opt.qpmin;
-            vc2index vc = vc2index(0, 0, _bse_ctotal);
-#pragma omp parallel for
-            for (int v1 = 0; v1 < _bse_vtotal; v1++) {
-                for (int c1 = 0; c1 < _bse_ctotal; c1++) {
-                    int index_vc = vc.I(v1, c1);
-                    // v->c
-                    for (int c2 = 0; c2 < _bse_ctotal; c2++) {
-                        int index_vc2 = vc.I(v1, c2);
-                        H(index_vc2, index_vc) += _Hqp(c2 + _bse_vtotal+offset, c1 + _bse_vtotal+offset);
-                    }
-                    // c-> v
-                    for (int v2 = 0; v2 < _bse_vtotal; v2++) {
-                        int index_vc2 = vc.I(v2, c1);
-                        H(index_vc2, index_vc) -= _Hqp(v2+offset, v1+offset);
-                    }
-                }
-            }
-            return;
-        }
+    elapsed_time = hend - hstart;
+    CTP_LOG(ctp::logDEBUG, _log)
+          << ctp::TimeStamp() << " Full matrix assembled in " 
+          << elapsed_time.count() << " secs" << flush;  
 
-template <typename T>
-    void BSE::Add_Hd(Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& H) {
-            int auxsize = _Mmn.auxsize();
-            vc2index vc = vc2index(0, 0, _bse_ctotal);
-            const int vmin=_opt.vmin-_opt.rpamin;
-            const int cmin=_bse_cmin-_opt.rpamin;
-#pragma omp parallel for
-            for (int v1 = 0; v1 < _bse_vtotal; v1++) {
-                const MatrixXfd Mmn1T = (_Mmn[v1 + vmin ].block(vmin, 0, _bse_vtotal, auxsize) * _epsilon_0_inv.asDiagonal()).transpose();
-                for (int c1 = 0; c1 < _bse_ctotal; c1++) {
-                    const MatrixXfd& Mmn2 = _Mmn[c1 + cmin];
-                    const MatrixXfd Mmn2xMmn1T = Mmn2.block(cmin, 0, _bse_ctotal, auxsize)*Mmn1T;
-                    int i1 = vc.I(v1, c1);
-                    for (int v2 = 0; v2 < _bse_vtotal; v2++) {
-                        for (int c2 = 0; c2 < _bse_ctotal; c2++) {
-                            int i2 = vc.I(v2, c2);
-                            H(i2, i1) -= Mmn2xMmn1T(c2, v2);
-                        }
-                    }
-                }
-            }
-
-            return;
-        }
+    hstart = std::chrono::system_clock::now();
+    tools::linalg_eigenvalues(hfull, energies, coefficients, _opt.nmax);
```

Comment:
if Davidson works fully and you are convinced, please remove this fully then.

Reply:
I would prefer keeping it just so that we can have a small test where we compare Davidson to DSYVEX.

Reply:
I might be overly careful here but I would like to run tests on larger systems, before removing the lapack diagonalization. Could we test a bit more and then remove it ?

Reply:
okay